Add CodeLimit workflow

robvanderleek opened the pull request:
## Overview

Thanks for this wonderful project.
Are you open to include a simple code quality check in the automated build system?
I'm developing an open source tool called [CodeLimit](https://getcodelimit.github.io) that monitors function length.
Still in active development, looking for any kind of feedback from the community.
The generated report for this codebase is attached below.

## CodeLimit Report

### Overview
| **Language** | **Files** | **Functions** | **Lines of Code** | **⚠** | **⛌** |
| --- | ---: | ---: | ---: | ---: | ---: |
| Java | 889 | 4768 | 25769 | 23 | 0 |
| JavaScript | 2 | 22 | 35 | 0 | 0 |
| **Totals** | **891** | **4790** | **25804** | **23** | **0** |

### Summary
| **Easy / Verbose** | **Hard-to-maintain ⚠** | **Unmaintainable ⛌** |
| ---: | ---: | ---: |
| 96% | 4% | 0% |

✅ 96% of the functions are maintainable, no refactoring necessary.

### Findings
| **Function** | **Length** | **File** |
| --- | ---: | --- |
| ⚠ [assertLinesMatchWithFastForward](https://github.com/robvanderleek/junit5/blob/HEAD/junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertLinesMatch.java#L115-L186) | 54 | junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertLinesMatch.java |
| ⚠ [resolve](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-engine/src/main/java/org/junit/platform/engine/support/discovery/EngineDiscoveryRequestResolution.java#L119-L166) | 48 | junit-platform-engine/src/main/java/org/junit/platform/engine/support/discovery/EngineDiscoveryRequestResolution.java |
| ⚠ [provideArguments](https://github.com/robvanderleek/junit5/blob/HEAD/junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/EmptyArgumentsProvider.java#L42-L93) | 48 | junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/EmptyArgumentsProvider.java |
| ⚠ [orderChildrenTestDescriptors](https://github.com/robvanderleek/junit5/blob/HEAD/junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/AbstractOrderingVisitor.java#L60-L132) | 47 | junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/AbstractOrderingVisitor.java |
| ⚠ [walk](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTreeWalker.java#L51-L100) | 46 | junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTreeWalker.java |
| ⚠ [nullSafeToString](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-commons/src/main/java/org/junit/platform/commons/util/StringUtils.java#L160-L205) | 42 | junit-platform-commons/src/main/java/org/junit/platform/commons/util/StringUtils.java |
| ⚠ [isWideningConversion](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java#L497-L550) | 42 | junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java |
| ⚠ [assertArrayElementsEqual](https://github.com/robvanderleek/junit5/blob/HEAD/junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertArrayEquals.java#L351-L392) | 41 | junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertArrayEquals.java |
| ⚠ [toStream](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-commons/src/main/java/org/junit/platform/commons/util/CollectionUtils.java#L189-L229) | 41 | junit-platform-commons/src/main/java/org/junit/platform/commons/util/CollectionUtils.java |
| ⚠ [addTestSource](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-reporting/src/main/java/org/junit/platform/reporting/open/xml/OpenTestReportGeneratingListener.java#L306-L346) | 41 | junit-platform-reporting/src/main/java/org/junit/platform/reporting/open/xml/OpenTestReportGeneratingListener.java |

13 more rows

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

marcphilipp commented:
Thanks for the initiative and the PR! We'll discuss whether to adopt this in our next team call.



> ❌ [apply](https://github.com/robvanderleek/junit5/blob/HEAD/junit-platform-commons/src/main/java/org/junit/platform/commons/function/Try.java?rgh-link-date=2024-12-24T09%3A12%3A57Z#L208-L288) 	61 	junit-platform-commons/src/main/java/org/junit/platform/commons/function/Try.java

@robvanderleek That is not actually a function body. Something seems to throw off your parser. :thinking:

marcphilipp commented:
Moreover, please rebase since I've improved `SuiteLauncherDiscoveryRequestBuilder` (based on your findings) on `main` in the meantime. :slightly_smiling_face:

robvanderleek commented:
Hi @marcphilipp 

Thank you for taking the time to look into CodeLimit.
Also, encouraging to see that a finding from the tool could be used to improve the code 🙂 

I've rebased the branch and updated the report (when the CodeLimit action runs inside the repository it will update the report automatically).
Now only 1 critical finding remains which is indeed a parse error from the tool, as you mentioned.

I'm currently using a parsing implementation that aims to be performant, robust, versatile (in terms of language support), but also accurate. I will look into the issue.

Again, thanks for the valuable feedback.

robvanderleek commented:
Hi @marcphilipp 

I've fixed the parse error and updated the report above.
There are no critical errors left, only 13 findings of methods that could become critical with further growth.

A CodeLimit badge for this repo would now look like this:

![image](https://raw.githubusercontent.com/robvanderleek/junit5/0b4adda97b7ef2efdd39778f4c99efb1767c8114/robvanderleek-add-codelimit-workflow/badge.svg)

Background on the meaning of the badge can be found [here in the documentation](https://getcodelimit.github.io/getting-started/development-workflow/#meaning-of-the-badge).

marcphilipp commented:
**Team decision:** We tend to review PRs and other code very carefully. The outliers listed above are long for good reasons. So, all in all, we don't want to add another tool to the build at this time.

robvanderleek commented:
Hi @marcphilipp 

Thanks for the feedback.

Since the codebase only has a few methods that exceed the 30 LOC limit (hard to maintain), and currently no unmaintainable methods (> 60 LOC), I'd also say the review process suffices.
OTOH, the `apply` method that was refactored slipped through the review process, and there is a risk the current outliers keep growing. I don't think any of the current build tools catches that.

I really appreciate the team's effort to evaluate CodeLimit.
All the best with this wonderful project 🙂
